<a href="https://colab.research.google.com/github/rgiuffre90/DS-Unit-1-Sprint-2-Statistics/blob/master/Robert_Giuffre_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
### YOUR WORK HERE

In [8]:
import pandas as pd
import numpy as np
from scipy import stats

In [9]:
column_headers = ['symbolizing','normalized-losses','make','fuel-type','aspiration','num-of-doors','body-style','drive-wheels','engine-location',
                 'wheel-base','legnth','width','height','curb-weight','engine-type','num-of-cylinders','engine-size','fuel-system','bore','stroke',
                 'compression-ratio','horsepower','peak-rpm','city-mpg','highway-mpg','price']

df= pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=column_headers)

df.head()

symbolizing normalized-losses         make  ... city-mpg highway-mpg  price
0            3                 ?  alfa-romero  ...       21          27  13495
1            3                 ?  alfa-romero  ...       21          27  16500
2            1                 ?  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [10]:
df= df.replace({'?' : np.NaN})

df.head()

symbolizing normalized-losses         make  ... city-mpg highway-mpg  price
0            3               NaN  alfa-romero  ...       21          27  13495
1            3               NaN  alfa-romero  ...       21          27  16500
2            1               NaN  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [11]:
df.describe()

symbolizing  wheel-base  ...    city-mpg  highway-mpg
count   205.000000  205.000000  ...  205.000000   205.000000
mean      0.834146   98.756585  ...   25.219512    30.751220
std       1.245307    6.021776  ...    6.542142     6.886443
min      -2.000000   86.600000  ...   13.000000    16.000000
25%       0.000000   94.500000  ...   19.000000    25.000000
50%       1.000000   97.000000  ...   24.000000    30.000000
75%       2.000000  102.400000  ...   30.000000    34.000000
max       3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
### YOUR WORK HERE

In [22]:
#create a sample of 10 cars 
sample = df.sample(10, random_state = 30)

sample.head()

symbolizing normalized-losses           make  ... city-mpg highway-mpg  price
121            1               154       plymouth  ...       31          38   6692
176           -1                65         toyota  ...       27          32  10898
24             1               148          dodge  ...       31          38   6229
88            -1               137     mitsubishi  ...       23          30   9279
72             3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

In [23]:
# Calculate mean of highway-mpg
df['highway-mpg'].mean()

30.75121951219512

In [26]:
#1samp t-test of highway-mpg
stats.ttest_1samp(sample['highway-mpg'], 30)

Ttest_1sampResult(statistic=1.218031060188056, pvalue=0.2541795875096462)

### Conclusion

Based on the T-statistic, we can conclude there is a difference in the mean highway-mpg compared to the sample size population mean.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
### YOUR WORK HERE

### Null Hypothesis:

Cars sold have a typical highway fuel efficiency of 35 miles per gallon.

### Alternative Hypothesis:

Cars sold do not have a typical highway fuel efficiency of 35 miles per gallon.

### Confidence Level 95%

In [27]:
sample = df.sample(10, random_state = 30)

stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

### Conclusion:

Based on a t-statistic of .957 and a p-value of .36355 we can **fail to reject** the null hypothesis that cars sold have a typical highway fuel efficiency of 35 miles per gallon.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE

In [28]:
sample2 = df.sample(100, random_state= 30)

stats.ttest_1samp(sample2['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### Conclusion:

The two tests results are different due to a larger sample size being used. 

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

*The results are different because the sample size is now 10x larger than the original sample size.*

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE

In [29]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-08-10 19:12:16--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-08-10 19:12:17 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [30]:
#create original dataframe

column_headers2 = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

In [31]:
df = pd.read_csv('house-votes-84.data', names=column_headers2)

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [32]:
#replace y, n, ?

df= df.replace({'y' : 1, 'n' : 0, "?" : np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [33]:
#create 2 seperate dataframes based on parties

rep = df[df['party'] == 'republican']
dem = df[df['party'] =='democrat']

dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [34]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### YOUR WORK HERE

In [40]:
# Locate issues based on support by democrats and republicans so t-tests can be run

dem.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [41]:
rep.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

### Null Hypothesis:

Democrats support for the 'anti satellite bill' is equal to Republican levels of support for the 'anti satellite bill'

### Alternative Hypothesis:

Democrats support for the 'anti satellite bill' is not equal to Republican levels of support for the 'anti satellite bill'

### Confidence Level 99%

In [37]:
stats.ttest_ind(dem['anti-satellite-ban'], rep['anti-satellite-ban'], nan_policy='omit')

Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

###Conclusion:
Based on a t-statistic of 12.526 and p-value of 8.52e-31 I **reject** the null hypothesis that democrats level of support for the 'anti satellite bill' is equal to republican levels of support for the 'anti satellite bill'

In [38]:
dem['anti-satellite-ban'].mean()

0.7722007722007722

In [39]:
rep['anti-satellite-ban'].mean()

0.24074074074074073

###Null Hypothesis
The democrats levels of support for the 'crime' bill is equal to the republican levels of support for the 'crime' bill

###Alternative Hypothesis
The democrats level of support for the 'crime' bill is not equal to the republican levels of support for the 'crime' bill

###Confidence Level 99%

In [47]:
stats.ttest_ind(dem['crime'], rep['crime'], nan_policy='omit')

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.952342705606092e-47)

### Conclusion:

Based on a t-statistic of -16.34 and p-value of 9.95e-31 I **reject** the null hypothesis that democrats level of support for the 'crime' is equal to republican levels of support for the 'crime' bill

In [48]:
dem['crime'].mean()

0.35019455252918286

In [49]:
rep['crime'].mean()

0.9813664596273292

### Null Hypothesis

The democrats level of support for the 'water-project' bill is equal to the republicans level of support for the 'water-project' bill

### Alternative Hypothesis

The democrats level of support for the 'water-project' bill is not equal to the republicans level of support for the 'water-project' bill


### Confidence Level 90%

In [45]:
stats.ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

### Conclusion

Based on a t-statistic of -.0889 and a p-value of .929 I **fail to reject** the null hypothesis that the democrats level of support for the 'water-project' bill is equal to the republicans level of support for the 'water-project' bill


## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE